File: DailyCodingProblems/398_deleteKthElemInList_Amazon/c/main.c
#include<stdio.h>
#include<stdlib.h>

typedef struct Node{
    int val;
    struct Node *next;
}Node,*node;

void insert(node* proot,int val);
node delete(node root,int k);
void visit(node root);

int main(){
    node root = NULL;
    insert(&root,1);
    insert(&root,2);
    insert(&root,3);
    insert(&root,4);
    insert(&root,5);
    root = delete(root,0);
    insert(&root,5);
    visit(root);
    return 0;
}

//遍历元素
void visit(node root){
    if(root == NULL) return;
    node tmp = NULL;
    for(tmp = root; tmp ; tmp = tmp->next){
        printf("%d%s",tmp->val,tmp->next ? "->" : "");
    }
    puts("");
}

//插入元素
void insert(node* proot,int val){
    node root = *proot;
    root = (node)malloc(sizeof(Node));
    root->val = val;
    root->next = *proot;
    *proot = root;
}

//这里假设k从0开始
node delete(node root,int k){
    node tmp = NULL;
    if(k == 0){ //删除第一个元素
        tmp = root->next;
        free(root);
        root = tmp;
    }else{  //删除其他元素
        int i;
        tmp = root;
        for(i = 1; i < k && tmp != NULL; i++){
            tmp = tmp->next;
        }
        if(tmp->next){
            node next = tmp->next;
            tmp->next = tmp->next->next;
            free(next);
        }
    }
    return root;
}

File: DailyCodingProblems/398_deleteKthElemInList_Amazon/python/main.py
#!/usr/bin/env python3
#coding:utf-8
# 数据节点类
class Node:
    next = None
    val = 0
    def __init__(self,val):
        self.val = val

# 链表类
class List:
    first = None
    length = 0
    def __init__(self):
        self.first = None
        self.length = 0
    # 插入新节点
    def insert(self,val):
        node = Node(val)
        node.next = self.first
        self.first = node
        self.length += 1
        return self
    # 删除指定位置的节点
    def delete(self,k):
        if self.length == 0: #list长度为0
            return False
        if k == 0:
            first = self.first.next
            del self.first
            self.first = first
        else:
            first = self.first
            for i in range(1,k):
                first = first.next
            if not first.next is None:
                next = first.next
                first.next = first.next.next
                del next
        return True
 
    def visit(self):
        first = self.first
        while not first is None:
            print(first.val,end = '')
            if not first.next is None:
                print('->',end = '')
            first = first.next
        print('')

li = List()
li.insert(1)
li.visit()
li.insert(3)
li.visit()
li.insert(4)
li.visit()
li.delete(1)
li.visit()
